src/appleseed/renderer/modeling/camera/projectpoints.cpp

```diff
@@ -0,0 +1,141 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "projectpoints.h"
+
+// appleseed.renderer headers.
+#include "renderer/modeling/entity/onframebeginrecorder.h"
+#include "renderer/modeling/entity/onrenderbeginrecorder.h"
+#include "renderer/modeling/frame/frame.h"
+#include "renderer/modeling/project/project.h"
+#include "renderer/modeling/scene/scene.h"
+#include "renderer/utility/paramarray.h"
+
+// appleseed.foundation headers.
+#include "foundation/utility/iostreamop.h"
+
+// Standard headers.
+#include <string>
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+struct ProjectPoints::Impl
+{
+    Camera*                     m_camera;
+    auto_release_ptr<Project>   m_project;
+    OnRenderBeginRecorder       m_render_begin_recorder;
+    OnFrameBeginRecorder        m_frame_begin_recorder;
+
+    Impl()
+      : m_camera(nullptr)
+    {
+    }
+};
+
+ProjectPoints::ProjectPoints(
+    auto_release_ptr<Camera> camera,
+    const Vector2u&          resolution)
+  : impl(new Impl())
+{
+    auto_release_ptr<Scene> scene = SceneFactory::create();
+
+    const string camera_name = camera->get_name();
+    scene->cameras().insert(camera);
+
+    impl->m_project = ProjectFactory::create("project_points");
+    impl->m_project->set_scene(scene);
+
+    impl->m_project->set_frame(
+        FrameFactory::create(
+            "frame",
+            ParamArray()
+                .insert("resolution", resolution)
+                .insert("camera", camera_name.c_str())));
+
+    bool success =
+        impl->m_project->get_scene()->on_render_begin(
+            impl->m_project.ref(),
+            nullptr,
+            impl->m_render_begin_recorder);
+
+    if (!success)
+        return;
+
+    success =
+        impl->m_project->get_scene()->on_frame_begin(
+            impl->m_project.ref(),
+            nullptr,
+            impl->m_frame_begin_recorder);
+
+    if (!success)
+        return;
+
+    impl->m_camera = impl->m_project->get_scene()->get_active_camera();
+}
+
+ProjectPoints::~ProjectPoints()
+{
+    impl->m_frame_begin_recorder.on_frame_end(impl->m_project.ref());
+    impl->m_render_begin_recorder.on_render_end(impl->m_project.ref());
+
+    delete impl;
+}
+
+bool ProjectPoints::is_initialized() const
+{
+    return impl->m_camera != nullptr;
```

Comment:
This works but it's slightly hacky to rely on the camera being set _last_ in the constructor. An explicit flag would be cleaner.